### DISCUSSION TITLE
sygus-stream doesn't produce multiple solutions

### DISCUSSION THREAD

**Discussion by lvtao-sec** [cvc5/cvc5#10439] [Q&A] [answered]
Hi all,

I am new to cvc5 and Sygus and meet the following question.
I would appreciate it if anyone could kindly help me.
Thanks in advance.

I wrote the following sygus code

```
(set-logic NIA)

(synth-fun equalSeq ((r1 Int)) Int
    ((I Int))
    ((I Int (r1 2
            (+ I I)
            (* I I)
            )))
)

(declare-var r1 Int)

(constraint (= (equalSeq r1) (+ r1 r1)))

(check-synth)
```

I expect it can at least give me two solutions like below:

```
(define-fun equalSeq ((r1 Int)) Int (+ r1 r1))
(define-fun equalSeq ((r1 Int)) Int (* r1 2))
```

However, it only gives me one at the very beginning and then continues exploring.
After adding the `-o sygus` option, I can see the enumeration process, which is shown below:

I'm just wondering why it skips (sygus-enum (* r1 2)) at the beginning.

```
(sygus-enum r1)
(sygus-candidate (equalSeq r1))
(sygus-enum 2)
(sygus-candidate (equalSeq 2))
(sygus-enum (+ r1 r1))
(sygus-candidate (equalSeq (+ r1 r1)))
(define-fun equalSeq ((r1 Int)) Int (+ r1 r1))
(sygus-enum (* r1 r1))
(sygus-enum (+ r1 2))
(sygus-enum (+ 2 2))
(sygus-enum (+ r1 (+ r1 r1)))
(sygus-enum (* r1 (+ r1 r1)))
(sygus-candidate (equalSeq (* r1 (+ r1 r1))))
(sygus-enum (+ r1 (* r1 r1)))
(sygus-enum (* r1 (* r1 r1)))
(sygus-enum (+ r1 (+ r1 2)))
(sygus-enum (* r1 (+ r1 2)))
(sygus-enum (+ r1 (+ 2 2)))
(sygus-enum (* r1 (+ 2 2)))
(sygus-enum (+ 2 (* r1 r1)))
(sygus-enum (* 2 (+ r1 2)))
(sygus-enum (+ 2 (+ 2 2)))
(sygus-enum (* 2 (+ 2 2)))
(sygus-enum (* r1 (+ r1 (+ r1 r1))))
(sygus-enum (+ r1 (* r1 (+ r1 r1))))
(sygus-enum (* r1 (* r1 (+ r1 r1))))
(sygus-candidate (equalSeq (* r1 (* r1 (+ r1 r1)))))
(sygus-enum (* r1 (+ r1 (* r1 r1))))
(sygus-enum (+ r1 (* r1 (* r1 r1))))
(sygus-enum (* r1 (* r1 (* r1 r1))))
(sygus-enum (+ r1 (+ r1 (+ r1 2))))
(sygus-enum (* r1 (+ r1 (+ r1 2))))
(sygus-enum (+ r1 (* r1 (+ r1 2))))
(sygus-enum (* r1 (* r1 (+ r1 2))))
```

**Comment by ajreynol** [comment, marked as answer]
`--sygus-stream` will enumerate solutions that are unique up to rewriting. In this case, `(* r1 2)` and `(+ r1 r1)` are equivalent via rewriting.

To make sygus enumerate *all* terms, you can use `--sygus-stream --sygus-rewriter=none`, which causes cvc5 to output:

```
(define-fun equalSeq ((r1 Int)) Int (+ r1 r1))
(define-fun equalSeq ((r1 Int)) Int (* r1 2))
(define-fun equalSeq ((r1 Int)) Int (* 2 r1))
```

However, in general, disabling the rewriter makes sygus enumeration much less efficient, since it considers all combinations of terms even those that are clearly equivalent.

**Comment by lvtao-sec** [comment]
Hi Andrew,

Thank you so much for the swift answer. It helps me a lot.

I have a few more questions here. I'd appreciate it if you could give me some clues.

Given two variables: r1 and r2
and the constraints:
r1 = r1 + 3
r2 == r1 + 5

I would expect the solver can give me something like: r1 = r1 + 2, r1 = r1 + 1, and r2 = r1 + 2
That means the syntax description should have assignments to and from variables.

Do you have any idea how we can implement this in SyGus?
Thank you so much!

**Comment by ajreynol** [comment]
We support multiple functions-to-synthesize. Functions-to-synthesize moreover can reference other previously declared functions in their grammar.  Here is an example: https://github.com/cvc5/cvc5/blob/main/test/regress/cli/regress1/sygus/fwd-decl-2.sy

If you have a particular example in mind that cvc5 does not handle, do let me know.

**Comment by lvtao-sec** [comment]
Hi Andrew,

Thanks again for your swift response.

I'm still struggling to find a way to solve my requirement by using SyGus/CVC5.
I'd appreciate it if you could shed some light on it.

The requirement is described below:
We have programs described in the imperative languages, let's say C language.
Given one piece of code written in C, we want to utilize SyGus and CVC5 to generate equivalent pieces of code.
Here we define the equivalence as the final variables values are exactly the same.

For example, given such a function `org-func`
```c
void org-func(int a, int b) {
    a = a + 2;
    b = a + 3;
    // Final state: a is original a plus 2, while b is original a plush 5
}
```
I would expect it to generate a new function, `new-func`, which is equivalent to `org-func`, like below:
```c
void new-func(int a, int b) {
    a = a + 2;
    b = a + 1;
    b = b + 2;
}
```

To realize the above requirement,
I need to 
(1) describe the syntax of the programming language using SyGus like below. **However, I have no idea how to express the assignment to a and b after each operation.**
```
(synth-fun equalSeq ((a Int) (b Int)) Int
    ((I Int))
    ((I Int (
            a b
            1 2 3 4 5 6
            (+ I I)
            ))
    )
)
```
(2) Express the final variable states as constraints. The constraints should be: the final values of a and b should be original a plus 2, and original a plus 5. **However, AFAIK, it looks like SyGus can only express the constraints on the single return value of the synthesized function.**

I'd really appreciate it if you can have time to answer my questions.
Thanks in advance.

**Comment by ajreynol** [comment]
You can model a function with multiple return values either as two functions-to-synthesize, or as a single function returning a `Tuple` type, see e.g. https://cvc5.github.io/docs/cvc5-1.1.2/theories/datatypes.html#tuples.

SyGuS has some limitations in modelling imperative code, there are more advanced alternatives for this e.g. https://www.semgus.org/.
